extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1026,9 +1027,34 @@ private List<SegmentIdWithShardSpec> generateSegmentIdsWithShardSpecsForAppend(
       previousSegmentId = allocation.asSegmentId().toString();
     }
 
+    final Map<DateTime, List<Pair<Integer, ClusterByPartition>>> partitionsByBucket = new HashMap<>();
+    for (int i = 0; i < partitionBoundaries.ranges().size(); i++) {
```

Comment:
We can ideally do all these checks in `https://github.com/apache/druid/blob/4e0ea7823b128cc214df8dc774d370b306b91461/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java#L912`

Reply:
> I guess we can skip this iteration entirely if the context value is null no ?

Done. I added this small optimization for the `INSERT` branch. `REPLACE` constructs the map .

> We can ideally do all these checks in https://github.com/apache/druid/blob/4e0ea7823b128cc214df8dc774d370b306b91461/extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java#L912

Yeah, but since `INSERT` doesn't really need the `partitionBoundaries` it didn't make a lot of sense move this to `generateSegmentIdsWithShardSpecs()`, but let me know if you think otherwise.

Reply:
I see. Thanks

Nit: 
Can we have a common method for this piece of code. Seems to be used both in insert and replace. 

```
  final Map<DateTime, List<Pair<Integer, ClusterByPartition>>> partitionsByBucket = new HashMap<>();
    for (int i = 0; i < partitionBoundaries.ranges().size(); i++) {
      ClusterByPartition partitionBoundary = partitionBoundaries.ranges().get(i);
      final DateTime bucketDateTime = getBucketDateTime(partitionBoundary, segmentGranularity, keyReader);
      partitionsByBucket.computeIfAbsent(bucketDateTime, ignored -> new ArrayList<>())
                        .add(Pair.of(i, partitionBoundary));
    }
    
    ``` 
    ?